parser/parser.go

```diff
@@ -0,0 +1,629 @@
+package parser
+
+import (
+	"strings"
+	"unicode"
+)
+
+// Message represents a parsed IRC message.
+type Message struct {
+	Tags    []Tag
+	Prefix  Prefix
+	Command string
+	Params  []string
+}
+
+// Error is
+type Error struct {
+	Partial Message
```

Comment:
What is the purpose of `Partial Message`?

Reply:
It's mainly just context for the state error. There are multiple ways to get to certain states, so having this info helps figure that out.